README.md
# pygments-lexer-solidity

Solidity and Yul lexer for Pygments, distributed as a PyPI package.

If you want pretty syntax highlighting in documentation for Solidity
files, and you\'re using Pygments, this might just be the thing for you.

Currently, Solidity keywords up to version 0.8.0 are included, to the
best of my ability. MRs are welcome!

The Yul intermediate language is also supported, including for
stand-alone code blocks.

## Overview

This repo builds on top of the source [which is can be located at https://gitlab.com/veox/pygments-lexer-solidity](https://gitlab.com/veox/pygments-lexer-solidity/).

To use the lexer for LaTeX highlighting simply extract the `lexer-latex.zip` file and upload both the `__init__.py` and `lexer.py` in your document directory or use a service like [overleaf.com](https://www.overleaf.com/).

A reference LaTeX document and generated PDF are included. This was rendered and generated using the overleaf web application.

## LaTeX

### Solidity

```latex
\begin{minted}{lexer.py:SolidityLexer -x}
   ... your code here ...
\end{minted}
```

### Yul

```latex
\begin{minted}{lexer.py:YulLexer -x}
   ... your code here ...
\end{minted}
```

## Installation

Set up your virtual environment, no matter how you do it.

``` shell
 virtualenv .virtualenv/`basename $(pwd)`
 source .virtualenv/`basename $(pwd)`/bin/activate
```

Then install via PyPI:

``` shell
 pip install pygments-lexer-solidity
```

Or from a local git repo:

```shell
 pip install -r requirements.txt
 pip install -e .
```

## Usage

Depends on doc-building infrastructure.

### Sphinx

Have this in Sphinx\'s `conf.py`:

```python
from sphinx.highlighting import lexers
from pygments_lexer_solidity import SolidityLexer, YulLexer
lexers['solidity'] = SolidityLexer()
lexers['yul'] = YulLexer()
```

Then use `.. code-block:: solidity` for Solidity code blocks, or
`.. code-block:: yul` for Yul.

### Command-line

To test a local copy of the lexer on the CLI:

``` shell
 pygmentize -x -l pygments_lexer_solidity/lexer.py:SolidityLexer example.sol
```

To generate a colorized HTML file:

```shell
 pygmentize -v -O full,style=fruity -o example.sol.html example.sol
```

## License

BSD 2-clause simplified. See `LICENSE.txt`.


CHANGELOG.rst
python-lexer-solidity change log
================================

[v0.7.0] - 2021-03-25
---------------------
Added
^^^^^
* Keywords from Solidity v0.7 and v0.8. (`#5`_)
* In-comment highlighting before the SPDX license identifier. (`#6`_)

Thanks to @axic!

.. _#5: https://gitlab.com/veox/pygments-lexer-solidity/-/merge_requests/5
.. _#6: https://gitlab.com/veox/pygments-lexer-solidity/-/merge_requests/6


[v0.6.0] - 2021-03-25
---------------------
Added
^^^^^
* Two new NatSpec keywords, ``@inheritdoc`` and ``@custom:..``.
* Python version requirement in ``setup.py``. (`#1`_)

.. _#1: https://gitlab.com/veox/pygments-lexer-solidity/-/issues/1


[v0.5.1] - 2020-04-20
---------------------
Changed
^^^^^^^
* Detect ``:`` as punctuation when used in Yul variable declaration
  with type hints.


[v0.5.0] - 2020-04-07
---------------------
Added
^^^^^
* ``YulLexer``, a new lexer that has some commonalities with
  ``SolidityLexer``, but is made specifically to highlight code
  (and object descriptions!) written in the Yul language.
* Also added some missing instructions in inline assembly.


[v0.4.1] - 2020-04-03
---------------------
Added
^^^^^
* Missing keywords, such as ``calldata``, as identified in
  `lexer PR for Rouge`_.
* Use of ``$`` in names.

.. _lexer PR for Rouge: https://github.com/rouge-ruby/rouge/pull/760

Changed
^^^^^^^
* Corrected M, N ranges for ``[u]fixed{M}x{N}`` types.
* Re-classified many keywords from ``Keywords.Reserved`` to just
  ``Keywords``, since they now have functionality, or were previosly
  erroneously mis-classified.


[v0.4.0] - 2019-12-23
---------------------
Changed
^^^^^^^
* Updated keywords to those of ``solc`` v0.6.0.


[v0.3.1] - 2018-05-24
---------------------
Changed
^^^^^^^
* Only match actual "nested" keywords (global variables), not all
  instances of their partials; e.g., ``msg.sender`` but not ``sender``.


[v0.3.0] - 2018-05-24
---------------------
Added
^^^^^
* NatSpec ``@special`` keywords.
* Integers in exponential notation.
* Floating-point numbers.


[v0.2.0] - 2018-04-17
---------------------
Changed
^^^^^^^
* Updated keywords to those of ``solc`` v0.4.22.


[v0.1.0] - 2018-03-08
---------------------
Changed
^^^^^^^
* Updated keywords to those of ``solc`` v0.4.21.


[v0.0.3] - 2017-10-31
---------------------
Added
^^^^^
* Minimal instructions in README.

Changed
^^^^^^^
* Updated keywords to those of ``solc`` v0.4.18.


[v0.0.2] - 2017-04-27
---------------------
Changed
^^^^^^^
* License: from MIT to BSD.

[v0.0.1] - 2017-04-27
---------------------
Added
^^^^^
* Ported ``pygments-main`` `PR #626`_ to this package.

.. _PR #626: https://bitbucket.org/birkenfeld/pygments-main/pull-requests/626/add-solidity-lexer


pygments_lexer_solidity/solidity.py
# -*- coding: utf-8 -*-
"""
    pygments.lexers.solidity
    ~~~~~~~~~~~~~~~~~~~~~~~~

    Lexer for the Solidity language.

    :copyright: Copyright 2006-2015 by the Pygments team, see AUTHORS.
    :license: BSD, see LICENSE for details.
"""

import re

from pygments.lexer import RegexLexer, include, bygroups, default, using, \
    this, words, combined
from pygments.token import Text, Comment, Operator, Keyword, Name, String, \
    Number, Punctuation, Other

__all__ = ['SolidityLexer']


class SolidityLexer(RegexLexer):
    """
    For Solidity source code.
    """

    name = 'Solidity'
    aliases = ['sol', 'solidity']
    filenames = ['*.sol']
    mimetypes = []

    flags = re.DOTALL | re.UNICODE | re.MULTILINE

    def type_names(prefix, sizerange):
        """
        Helper for type name generation, like: bytes1 .. bytes32
        """
        namelist = []
        for i in sizerange: namelist.append(prefix + str(i))
        return tuple(namelist)

    def type_names_mn(prefix, sizerangem, sizerangen):
        """
        Helper for type name generation, like: fixed0x8 .. fixed0x256
        """
        lm = []
        ln = []
        namelist = []

        # construct lists out of ranges
        for i in sizerangem: lm.append(i)
        for i in sizerangen: ln.append(i)

        # sizes (in bits) are valid if (%8 == 0) and (m+n <= 256)
        # first condition is covered by passing proper sizerange{m,n}
        validpairs = [tuple([m,n]) for m in lm for n in ln if m+n<=256]
        for i in validpairs:
            namelist.append(prefix + str(i[0]) + 'x' + str(i[1]))

        return tuple(namelist)


    tokens = {
        'assembly': [
            include('comments'),
            include('numbers'),
            include('strings'),
            include('whitespace'),

            (r'\{', Punctuation, '#push'),
            (r'\}', Punctuation, '#pop'),
            (r'[(),]', Punctuation),
            (r':=|=:', Operator),
            (r'(let)(\s*)(\w*\b)', bygroups(Operator.Word, Text, Name.Variable)),
            (r'(\w*\b)(\:[^=])', bygroups(Name.Label, Punctuation)),

            (r'(stop|add|mul|sub|div|sdiv|mod|smod|addmod|mulmod|exp|'
             r'signextend|lt|gt|slt|sgt|eq|iszero|and|or|xor|not|byte|'
             r'sha3|address|balance|origin|caller|callvalue|calldataload|'
             r'calldatasize|calldatacopy|codesize|codecopy|gasprice|'
             r'extcodesize|extcodecopy|blockhash|coinbase|timestamp|'
             r'number|difficulty|gaslimit|pop|mload|mstore|mstore8|sload|'
             r'sstore|jump|jumpi|pc|msize|gas|jumpdest|push1|push2|'
             r'push32|dup1|dup2|dup16|swap1|swap2|swap16|log0|log1|log4|'
             r'create|call|callcode|return|delegatecall|suicide)\b',
             Name.Function),

            # everything else is either a local/external var, or label
            ('[a-zA-Z_]\w*', Name)
        ],
        'comments': [
            (r'//([\w\W]*?\n)', Comment.Single),
            (r'/[*][\w\W]*?[*]/', Comment.Multiline),
            # Open until EOF, so no ending delimiter
            (r'/[*][\w\W]*', Comment.Multiline),
        ],
        'keywords-other': [
            (words(('for', 'in', 'while', 'do', 'break', 'return',
                    'returns', 'continue', 'if', 'else', 'throw',
                    'new', 'delete'),
                   suffix=r'\b'), Keyword),

            (r'assembly\b', Keyword, 'assembly'),

            (words(('contract', 'enum', 'event', 'function',
                    'library', 'mapping', 'modifier', 'struct', 'var'),
                   suffix=r'\b'), Keyword.Declaration),

            (r'(import|using)\b', Keyword.Namespace),

            # misc keywords
            (r'pragma solidity\b', Keyword.Reserved),
            (r'(_|as|constant|default|from|is)\b', Keyword.Reserved),
            # built-in modifier
            (r'payable\b', Keyword.Reserved),
            # variable location specifiers
            (r'(memory|storage)\b', Keyword.Reserved),
            # method visibility specifiers
            (r'(external|internal|private|public)\b', Keyword.Reserved),
            # event parameter specifiers
            (r'(anonymous|indexed)\b', Keyword.Reserved),
            # added in `solc v0.4.0`, not covered elsewhere
            (r'(abstract|interface|pure|static|view)\b', Keyword.Reserved),

            (r'(true|false)\b', Keyword.Constant),
            (r'(wei|finney|szabo|ether)\b', Keyword.Constant),
            (r'(seconds|minutes|hours|days|weeks|years)\b', Keyword.Constant),
        ],
        'keywords-types': [
            (words(('address', 'bool', 'byte', 'bytes', 'int', 'fixed',
                    'string', 'ufixed', 'uint'),
                   suffix=r'\b'), Keyword.Type),

            (words(type_names('int', range(8, 256+1, 8)),
                   suffix=r'\b'), Keyword.Type),
            (words(type_names('uint', range(8, 256+1, 8)),
                   suffix=r'\b'), Keyword.Type),
            (words(type_names('bytes', range(1, 32+1)),
                   suffix=r'\b'), Keyword.Type),
            (words(type_names_mn('fixed', range(0, 256+1, 8), range(8, 256+1, 8)),
                   suffix=r'\b'), Keyword.Type),
            (words(type_names_mn('ufixed', range(0, 256+1, 8), range(8, 256+1, 8)),
                   suffix=r'\b'), Keyword.Type),
        ],
        'numbers': [
            (r'0[xX][0-9a-fA-F]+', Number.Hex),
            (r'[0-9]+', Number.Integer),
        ],
        'string-parse-common': [
            # escapes
            (r'\\(u[0-9a-fA-F]{4}|x..|[^x])', String.Escape),
            # almost everything else is plain characters
            (r'[^\\"\'\n]+', String),
            # line continuation
            (r'\\\n', String),
            # stray backslash
            (r'\\', String)
        ],
        'string-parse-double': [
            (r'"', String, '#pop'),
            (r"'", String)
        ],
        'string-parse-single': [
            (r"'", String, '#pop'),
            (r'"', String)
        ],
        'strings': [
            # hexadecimal string literals
            (r"hex'[0-9a-fA-F]+'", String),
            (r'hex"[0-9a-fA-F]+"', String),
            # usual strings
            (r'"', String, combined('string-parse-common',
                                    'string-parse-double')),
            (r"'", String, combined('string-parse-common',
                                    'string-parse-single'))
        ],
        'whitespace': [
            (r'\s+', Text)
        ],
        'root': [
            include('comments'),
            include('keywords-types'),
            include('keywords-other'),
            include('numbers'),
            include('strings'),
            include('whitespace'),

            (r'\+\+|--|\*\*|\?|:|~|&&|\|\||=>|==?|!=?|'
             r'(<<|>>>?|[-<>+*%&|^/])=?', Operator),

            (r'[{(\[;,]', Punctuation),
            (r'[})\].]', Punctuation),

            (r'(block|msg|now|this|super|tx)\b', Name.Builtin),
            # built-in members, should these be Name.Function?..
            (r'(selfdestruct|suicide)\b', Name.Builtin),
            (r'(balance|send)\b', Name.Builtin),
            (r'(call|callcode|delegatecall)\b', Name.Builtin),

            (r'(addmod|ecrecover|mulmod|ripemd160|sha256|sha3)\b', Name.Function),

            # everything else is a var/function name
            ('[a-zA-Z_]\w*', Name)
        ] # 'root'
    } # tokens


pygments_lexer_solidity/lexer.py
# -*- coding: utf-8 -*-
"""
    pygments.lexers.solidity
    ~~~~~~~~~~~~~~~~~~~~~~~~

    Lexer for the Solidity language.

    :copyright: Copyright 2006-2015 by the Pygments team, see AUTHORS.
    :license: BSD, see LICENSE for details.
"""

import re

from pygments.lexer import (
    RegexLexer,
    bygroups,
    combined,
    default,
    include,
    inherit,
    this,
    using,
    words,
)
from pygments.token import Text, Comment, Operator, Keyword, Name, String, \
    Number, Punctuation, Other

__all__ = ['SolidityLexer', 'YulLexer']


class BaseLexer(RegexLexer):
    """Common for both Solidity and Yul."""

    tokens = {
        'assembly': [
            include('comments'),
            include('numbers'),
            include('strings'),
            include('whitespace'),

            (r'\{', Punctuation, '#push'),
            (r'\}', Punctuation, '#pop'),
            (r'[(),]', Punctuation),
            (r':=|=:', Operator),
            (r'(let)(\s*)(\w*\b)',
             bygroups(Operator.Word, Text, Name.Variable)),

            # evm instructions; ordered by opcode
            (r'(stop|add|mul|sub|div|sdiv|mod|smod|addmod|mulmod|exp|'
             r'signextend|'
             r'lt|gt|slt|sgt|eq|iszero|and|or|xor|not|byte|shl|shr|sar|'
             r'keccak256|'
             r'address|balance|origin|caller|'
             r'callvalue|calldataload|calldatasize|calldatacopy|'
             r'codesize|codecopy|gasprice|extcodesize|extcodecopy|'
             r'returndatasize|returndatacopy|extcodehash|'
             r'blockhash|coinbase|timestamp|number|difficulty|gaslimit|'
             r'chainid|selfbalance|'
             r'pop|mload|mstore|mstore8|sload|sstore|'
             r'pc|msize|gas|'
             r'log0|log1|log2|log3|log4|'
             r'create|call|callcode|return|delegatecall|create2|'
             r'staticcall|revert|'
             r'invalid|selfdestruct)\b',
             Name.Function),

            # obsolete aliases for keccak256 and selfdestruct
            (r'(sha3|suicide)\b', Name.Function),

            (r'(case|default|for|if|switch)\b', Keyword),

            # everything else is either a local/external var, or label
            ('[a-zA-Z_]\w*', Name)
        ],
        'comment-parse-single': [
            include('natspec'),
            (r'\n', Comment.Single, '#pop'),
            (r'[^\n]', Comment.Single),
        ],
        'comment-parse-multi': [
            include('natspec'),
            (r'[^*/]', Comment.Multiline),
            (r'\*/', Comment.Multiline, '#pop'),
            (r'[*/]', Comment.Multiline),
        ],
        'comments': [
            (r'//', Comment.Single, 'comment-parse-single'),
            (r'/[*]', Comment.Multiline, 'comment-parse-multi'),
        ],
        'natspec': [
            (r'@(author|dev|inheritdoc|notice|param|return|title|'
             r'custom:[a-z][a-z-]*)\b',
             Comment.Special),
        ],
        'numbers': [
            (r'0[xX][0-9a-fA-F]+', Number.Hex),
            (r'[0-9][0-9]*\.[0-9]+([eE][0-9]+)?', Number.Float),
            (r'[0-9]+([eE][0-9]+)?', Number.Integer),
        ],
        'string-parse-common': [
            # escapes
            (r'\\(u[0-9a-fA-F]{4}|x..|[^x])', String.Escape),
            # almost everything else is plain characters
            (r'[^\\"\'\n]+', String),
            # line continuation
            (r'\\\n', String),
            # stray backslash
            (r'\\', String)
        ],
        'string-parse-double': [
            (r'"', String, '#pop'),
            (r"'", String)
        ],
        'string-parse-single': [
            (r"'", String, '#pop'),
            (r'"', String)
        ],
        'strings': [
            # hexadecimal string literals
            (r"hex'[0-9a-fA-F]+'", String),
            (r'hex"[0-9a-fA-F]+"', String),
            # unicode string literals
            (r'unicode"', String, combined('string-parse-common',
                                    'string-parse-double')),
            (r"unicode'", String, combined('string-parse-common',
                                    'string-parse-single')),
            # usual strings
            (r'"', String, combined('string-parse-common',
                                    'string-parse-double')),
            (r"'", String, combined('string-parse-common',
                                    'string-parse-single'))
        ],
        'whitespace': [
            (r'\s+', Text)
        ],
    } # tokens


class SolidityLexer(BaseLexer):
    """For Solidity source code."""

    name = 'Solidity'
    aliases = ['sol', 'solidity']
    filenames = ['*.sol', '*.solidity']
    mimetypes = ['text/x-solidity']

    flags = re.DOTALL | re.UNICODE | re.MULTILINE

    def type_names(prefix, sizerange):
        """
        Helper for type name generation, like: bytes1 .. bytes32

        Assumes that the size range passed in is valid.
        """
        namelist = []
        for i in sizerange: namelist.append(prefix + str(i))
        return tuple(namelist)

    def type_names_mn(prefix, sizerangem, sizerangen):
        """
        Helper for type name generation, like: fixed8x0 .. fixed26x80

        Assumes that the size range passed in is valid.
        """
        lm = []
        ln = []
        namelist = []

        # construct lists out of ranges
        for i in sizerangem: lm.append(i)
        for i in sizerangen: ln.append(i)

        validpairs = [tuple([m,n]) for m in lm for n in ln]
        for i in validpairs:
            namelist.append(prefix + str(i[0]) + 'x' + str(i[1]))

        return tuple(namelist)


    tokens = {
        'assembly': [
            # labels
            (r'(\w*\b)(\:[^=])',
             bygroups(Name.Label, Punctuation)),

            # evm instructions present in `assembly` but not Yul
            (r'(jump|jumpi|jumpdest)\b', Name.Function),
            # TODO: rename helper: instructions, not types
            (words(type_names('dup', range(1, 16+1)),
                   suffix=r'\b'), Keyword.Function),
            (words(type_names('swap', range(1, 16+1)),
                   suffix=r'\b'), Keyword.Function),
            (words(type_names('push', range(1, 32+1)),
                   suffix=r'\b'), Keyword.Function),

            inherit,
        ],
        'keywords-builtins': [
            # compiler built-ins
            (r'(balance|now)\b', Name.Builtin),
            (r'selector\b', Name.Builtin),
            (r'(super|this)\b', Name.Builtin),
        ],
        'keywords-functions': [
            # receive/fallback functions
            (r'(receive|fallback)\b', Keyword.Function),

            # like block.hash and msg.gas in `keywords-nested`
            (r'(blockhash|gasleft)\b', Name.Function),

            # single-instruction yet function syntax
            (r'(selfdestruct|suicide)\b', Name.Function),

            # processed into many-instructions
            (r'(send|transfer|call|callcode|delegatecall)\b',
             Name.Function),
            (r'(assert|revert|require)\b', Name.Function),
            (r'push\b', Name.Function),

            # built-in functions and/or precompiles
            (words(('addmod', 'ecrecover', 'keccak256', 'mulmod',
                    'sha256', 'sha3', 'ripemd160'),
                   suffix=r'\b'), Name.Function),
        ],
        'keywords-types': [
            (words(('address', 'bool', 'byte', 'bytes', 'int', 'string',
                    'uint'),
                   suffix=r'\b'), Keyword.Type),
            (words(type_names('bytes', range(1, 32+1)),
                   suffix=r'\b'), Keyword.Type),
            (words(type_names('int', range(8, 256+1, 8)),
                   suffix=r'\b'), Keyword.Type),
            (words(type_names('uint', range(8, 256+1, 8)),
                   suffix=r'\b'), Keyword.Type),

            # not yet fully implemented, therefore reserved types
            (words(('fixed', 'ufixed'), suffix=r'\b'), Keyword.Reserved),
            (words(type_names_mn('fixed',
                                 range(8, 256+1, 8),
                                 range(0, 80+1, 1)),
                   suffix=r'\b'), Keyword.Reserved),
            (words(type_names_mn('ufixed',
                                 range(8, 256+1, 8),
                                 range(0, 80+1, 1)),
                   suffix=r'\b'), Keyword.Reserved),
        ],
        'keywords-nested': [
            (r'abi\.encode(|Packed|WithSelector|WithSignature)\b',
             Name.Builtin),
            (r'block\.(blockhash|chainid|coinbase|difficulty|gaslimit|hash|'
             r'number|timestamp)\b', Name.Builtin),
            (r'msg\.(data|gas|sender|value)\b', Name.Builtin),
            (r'tx\.(gasprice|origin)\b', Name.Builtin),
        ],
        'keywords-other': [
            (words(('for', 'in', 'while', 'do', 'break', 'return',
                    'returns', 'continue', 'if', 'else', 'throw',
                    'new', 'delete', 'try', 'catch'),
                   suffix=r'\b'), Keyword),

            (r'assembly\b', Keyword, 'assembly'),

            (words(('contract', 'interface', 'enum', 'event', 'function',
                    'constructor', 'library', 'mapping', 'modifier',
                    'struct', 'var'),
                   suffix=r'\b'), Keyword.Declaration),

            (r'(import|using)\b', Keyword.Namespace),

            # pragmas are not pragmatic in their formatting :/
            (r'pragma( experimental| solidity| abicoder|)\b', Keyword),
            # misc keywords
            (r'(_|as|constant|from|is)\b', Keyword),
            (r'emit\b', Keyword),
            # built-in modifier
            (r'payable\b', Keyword),
            # variable location specifiers
            (r'(calldata|memory|storage)\b', Keyword),
            # method visibility specifiers
            (r'(external|internal|private|public)\b', Keyword),
            # event parameter specifiers
            (r'(anonymous|indexed)\b', Keyword),
            # added in solc v0.4.0, not covered elsewhere
            (r'(abstract|pure|static|view)\b', Keyword),
            # added in solc v0.6.0, not covered elsewhere
            (r'(override|virtual)\b', Keyword),
            # added in solc v0.8.0
            (r'(unchecked)\b', Keyword),
            # type information
            (r'type\(.*\)\.(min|max|interfaceId|creationCode|runtimeCode|name)\b', Keyword),

            # reserved for future use since don't-remember-when
            (words(('after', 'case', 'default', 'final', 'in', 'inline',
                    'let', 'match', 'null', 'of', 'relocatable', 'static',
                    'switch', 'typeof'),
                   suffix=r'\b'), Keyword.Reserved),
            # reserved for future use since solc v0.5.0
            (words(('alias', 'apply', 'auto', 'copyof', 'define',
                    'immutable', 'implements', 'macro', 'mutable',
                    'partial', 'promise', 'reference',
                    'sealed', 'sizeof', 'supports', 'typedef'),
                   suffix=r'\b'), Keyword.Reserved),

            # built-in constants
            (r'(true|false)\b', Keyword.Constant),
            (r'(wei|finney|szabo|ether)\b', Keyword.Constant),
            (r'(seconds|minutes|hours|days|weeks|years)\b',
             Keyword.Constant),
        ],

        'root': [
            inherit,
            include('comments'),
            include('keywords-builtins'),
            include('keywords-functions'),
            include('keywords-types'),
            include('keywords-nested'),
            include('keywords-other'),
            include('numbers'),
            include('strings'),
            include('whitespace'),

            (r'\+\+|--|\*\*|\?|:|~|&&|\|\||=>|==?|!=?|'
             r'(<<|>>>?|[-<>+*%&|^/])=?', Operator),

            (r'[{(\[;,]', Punctuation),
            (r'[})\].]', Punctuation),

            # everything else is a var/function name
            ('[a-zA-Z$_]\w*', Name),
        ],
    } # tokens


class YulLexer(BaseLexer):
    """For Yul stand-alone source code."""

    name = 'Yul'
    aliases = ['yul']
    filenames = ['*.yul']
    mimetypes = ['text/x-yul']

    flags = re.DOTALL | re.UNICODE | re.MULTILINE

    tokens = {
        'root': [
            inherit,

            # Yul-specific
            (r'(code|data|function|object)\b', Keyword),
            (r'data(copy|offset|size)\b', Keyword.Function),
            (r'->', Operator),

            include('assembly'),

            # ':' - variable declaration type hint - catch it last
            (':', Punctuation),
        ],
    } # tokens
